repo.js

```diff
@@ -1,81 +1,200 @@
+        this.url = url
       })
+    }
+
+    this._open()
+  }
+
+  // Attempt to create a websocket connection to a gateway if possible
+  _createWebsocket () {
+    if(!this.opts.gateway) return
+    const servers = [].concat(this.opts.gateway)
+    const server = chooseRandom(servers)
+
+    const url = setSecure(server + '/' + this.archive.key.toString('hex'))
+
+    this.websocket = websocket(url)
+
+    this.websocket.once('error', () => {
+      setTimeout(() => {
+        this._createWebsocket(server)
+      }, DEFAULT_WEBSOCKET_RECONNECT)
     })
-  })
+
+    this._replicate(this.websocket)
+  }
+
+  _startWebsocketTimer () {
+    // Wait a second before trying to establish a connection to the gateway in case there's already WebRTC peers
+    this._websocketTimer = setTimeout(() => {
+      this._createWebsocket()
+    }, DEFAULT_WEBSOCKET_CONNECTION_DELAY)
+  }
+
+  _joinWebrtcSwarm () {
+    const hubs = [].concat(this.opts.signalhub || DEFAULT_SIGNALHUBS).map(setSecure)
+
+    const appName = this.archive.discoveryKey.toString('hex').slice(40)
+
+    const signalhub = Signalhub(appName, hubs)
+
+    this.signalhub = signalhub
+
+    // Listen for incoming connections
+    const subscription = signalhub.subscribe('all')
+    const processSubscription = through.obj((data, enc, cb) => {
+      if (data.from === swarm.me) {
+        return cb()
+      }
+      if (data.type === 'connect') {
+        // If we've gotten a connection request, delay websocket connection
+        // This is to prioritize WebRTC traffic and reduce gateway load
+        clearInterval(this._websocketTimer)
+        this._websocketTimer = setTimeout(() => {
+          this._createWebsocket()
+        }, DEFAULT_WEBSOCKET_CONNECTION_DELAY_LONG)
+        cb()
+        // We did the thing so no need to listen any further
+        subscription.destroy()
+      }
+    })
+
+    subscription.pipe(processSubscription)
+
+    const swarm = new WebrtcSwarm(signalhub)
+
+    this.swarm = swarm
+
+    swarm.on('peer', (stream) => this._replicate(stream))
+  }
+
+  _replicate (stream) {
+    pump(stream, this.archive.replicate({
+      sparse: true,
+      live: true
+    }), stream)
+  }
+
+  _open () {
+    this.archive.ready(() => {
+      this._joinWebrtcSwarm()
+      this._startWebsocketTimer()
+      this._isReady = true
+      this.emit('ready')
+    })
+  }
+
+  ready (cb) {
+    if(this._isReady) {
+      setTimeout(cb, 0)
+    }
+    this.once('ready', cb)
+  }
+
+  close (cb) {
+    if(cb) this.once('close', cb)
+
+    // Close the gateway socket if one exists
+    if (this.websocket) {
+      this.websocket.end()
+      this.websocket = null
+    }
+
+    // Stop accepting new WebRTC peers
+    this.swarm.close(() => {
+      // Disconnect from the signalhub
+      this.signalhub.close(() => {
+        // Close the DB files being used by hyperdrive
+        this.archive.close(() => {
+          this.emit('close')
```

Comment:
do all of these properly call their respective callbacks? sometimes these things can be tricky